### PR TITLE
sdbus-cpp: Add API to get message credentials

### PR DESCRIPTION
**Pull request by alivenets**
Signed-off-by: Alexander Livenets <a.livenets@gmail.com>

### CODE REVIEW

**Comment by sangelovic** [comment]
Thanks for the contribution @alivenets. I'd a have a number of comments. Do you prefer that I place them and you'd modify the code and I re-review, or may I do the changes directly in your branch and merge?

File: src/SdBus.cpp [hunk 1]
```diff
@@ -260,4 +260,49 @@ sd_bus* SdBus::sd_bus_flush_close_unref(sd_bus *bus)
     return ::sd_bus_flush_close_unref(bus);
 }
 
+int SdBus::sd_bus_query_sender_creds(sd_bus_message *m, uint64_t mask, sd_bus_creds **c)
+{
+    return ::sd_bus_query_sender_creds(m, mask, c);
+}
```
Comment (sangelovic) [inline src/SdBus.cpp:266]: Would you please decorate all these new methods with `std::lock_guard lock(sdbusMutex_);`, just like (most of the) sd-bus functions above in the file? These sd-bus calls must be thread-safe.

File: src/Message.cpp [hunk 2]
```diff
@@ -616,6 +616,120 @@ bool Message::isEmpty() const
...
+pid_t Message::getCredsPid() const
+{
+    uint64_t mask = SD_BUS_CREDS_PID | SD_BUS_CREDS_AUGMENT;
+    sd_bus_creds *creds = nullptr;
+    int r = sdbus_->sd_bus_query_sender_creds((sd_bus_message*)msg_, mask, &creds);
+    SDBUS_THROW_ERROR_IF(r < 0, "Failed to get bus creds", -r);
+
+    pid_t pid = 0;
+    r = sdbus_->sd_bus_creds_get_pid(creds, &pid);
+    sd_bus_creds_unref(creds);
```
Comment (sangelovic) [inline src/Message.cpp:628]: I'd suggest to place unreffing into a `SCOPE_EXIT` macro in all these new methods, and place it right after `sd_bus_query_sender_creds()`. See a comment below explaining more on the macro.

**Review by sangelovic — CHANGES_REQUESTED** [review]

File: src/Message.cpp (hunk 2 above)
Comment (alivenets) [reply src/Message.cpp]: Thanks for the tip, fixed.

File: src/SdBus.cpp (hunk 1 above)
Comment (alivenets) [reply src/SdBus.cpp:268]: Thanks for the tip, fixed.

File: src/Message.cpp [hunk 3]
```diff
@@ -616,6 +616,120 @@ bool Message::isEmpty() const
...
+
+    const char *cLabel = nullptr;
+    r = sdbus_->sd_bus_creds_get_selinux_context(creds, &cLabel);
+    if (r < 0) {
+        sdbus_->sd_bus_creds_unref(creds);
+        SDBUS_THROW_ERROR("Failed to get bus cred selinux context", -r);
+    }
+    std::string label = cLabel;
+    sdbus_->sd_bus_creds_unref(creds);
+    return label;
```
[the remark this answers is not in the document]
Comment (alivenets) [reply src/Message.cpp]: Thanks for the comment, fixed.

File: src/Message.cpp [hunk 4]
```diff
@@ -616,6 +616,120 @@ bool Message::isEmpty() const
...
+    uint64_t mask = SD_BUS_CREDS_AUGMENT | SD_BUS_CREDS_SELINUX_CONTEXT;
+    sd_bus_creds *creds = nullptr;
+    int r = sdbus_->sd_bus_query_sender_creds((sd_bus_message*)msg_, mask, &creds);
+    SDBUS_THROW_ERROR_IF(r < 0, "Failed to get bus creds", -r);
+
+    const char *cLabel = nullptr;
+    r = sdbus_->sd_bus_creds_get_selinux_context(creds, &cLabel);
+    if (r < 0) {
+        sdbus_->sd_bus_creds_unref(creds);
+        SDBUS_THROW_ERROR("Failed to get bus cred selinux context", -r);
```
[the remark this answers is not in the document]
Comment (alivenets) [reply src/Message.cpp]: Fixed.

File: src/Message.cpp [hunk 5]
```diff
@@ -616,6 +616,120 @@ bool Message::isEmpty() const
...
+{
+    std::vector<gid_t> gidv{};
+    uint64_t mask = SD_BUS_CREDS_SUPPLEMENTARY_GIDS | SD_BUS_CREDS_AUGMENT;
+    sd_bus_creds *creds = nullptr;
+    int r = sdbus_->sd_bus_query_sender_creds((sd_bus_message*)msg_, mask, &creds);
+    SDBUS_THROW_ERROR_IF(r < 0, "Failed to get bus creds", -r);
+
+    const gid_t *gids = nullptr;
+    int num_gids = 0;
+    num_gids = sdbus_->sd_bus_creds_get_supplementary_gids(creds, &gids);
```
[the remark this answers is not in the document]
Comment (alivenets) [reply src/Message.cpp]: reused `r` for this purpose

File: src/Message.cpp [hunk 6]
```diff
@@ -616,6 +616,120 @@ bool Message::isEmpty() const
...
+    gid_t egid = 65534;
+    r = sdbus_->sd_bus_creds_get_egid(creds, &egid);
+    sdbus_->sd_bus_creds_unref(creds);
+    SDBUS_THROW_ERROR_IF(r < 0, "Failed to get bus cred egid", -r);
+    return egid;
+}
+
+std::vector<gid_t> Message::getCredsSupplementaryGids() const
+{
+    std::vector<gid_t> gidv{};
```
[the remark this answers is not in the document]
Comment (alivenets) [reply src/Message.cpp]: Renamed.

File: src/Message.cpp [hunk 7]
```diff
@@ -616,6 +616,120 @@ bool Message::isEmpty() const
...
+}
+
+uid_t Message::getCredsUid() const
+{
+    uint64_t mask = SD_BUS_CREDS_UID | SD_BUS_CREDS_AUGMENT;
+    sd_bus_creds *creds = nullptr;
+    int r = sdbus_->sd_bus_query_sender_creds((sd_bus_message*)msg_, mask, &creds);
+    SDBUS_THROW_ERROR_IF(r < 0, "Failed to get bus creds", -r);
+
+    uid_t uid = 65534;
```
[the remark this answers is not in the document]
Comment (alivenets) [reply src/Message.cpp]: This is usually uid for `nobody`. However, here it is not important. For clarity, used `(uid_t)-1`.

File: include/sdbus-c++/Message.h [hunk 8]
```diff
@@ -37,6 +37,7 @@
 #include <cstdint>
 #include <cassert>
 #include <functional>
+#include <vector>
```
[the remark this answers is not in the document]
Comment (alivenets) [reply include/sdbus-c++/Message.h]: Fixed

[the remark this answers is not in the document]
Comment (sangelovic) [reply include/sdbus-c++/Message.h]: Would you please add `#include "sys/types.h"`? This one defines `gid_t` and other types you've introduced...

**Review by sangelovic — APPROVED** [review]
Thank you!

**Comment by alivenets** [comment]
Thank you @sangelovic for a small fixup with `#include`.

**Pull request merged**